pkg/vpcmodel/commonConnectivity.go

```diff
@@ -6,7 +6,114 @@ SPDX-License-Identifier: Apache-2.0
 
 package vpcmodel
 
-import "github.com/np-guard/models/pkg/connection"
+import (
+	"github.com/np-guard/models/pkg/connection"
+	"github.com/np-guard/models/pkg/netp"
+)
+
+// todo: remove stateful from connection.Set (for both options)
+
+// ExtendedSet connection details
+type ExtendedSet struct {
+	statefulConn    *connection.Set // stateful TCP connection between <src, dst>
+	nonStatefulConn *connection.Set // nonstateful TCP connection between <src, dst>; complementary of statefulConn
+	otherConn       *connection.Set // non TCP connection (for which stateful is non-relevant)
+	conn            *connection.Set // entire connection
+}
+
+// operation on ExtendedSet
+// The operations are performed on the disjoint statefulConn and otherConn and on conn which contains them;
+// nonStatefulConn - the tcp complementary of statefulConn w.r.t. conn -
+// is computed as conn minus (statefulConn union otherConn)
+
+func NoConnsExtendedSet() *ExtendedSet {
+	return &ExtendedSet{
+		statefulConn:    NoConns(),
+		nonStatefulConn: NoConns(),
+		otherConn:       NoConns(),
+		conn:            NoConns(),
+	}
+}
+
+var all = connection.All()
+
+func (e *ExtendedSet) IsAllObliviousStateful() bool {
+	return e.conn.Equal(all)
+}
+
+func (e *ExtendedSet) IsEmpty() bool {
+	return e.conn.IsEmpty()
+}
+
+func (e *ExtendedSet) Equal(other *ExtendedSet) bool {
+	return e.statefulConn.Equal(other.statefulConn) && e.otherConn.Equal(other.otherConn) &&
+		e.conn.Equal(other.conn)
+}
+
+func (e *ExtendedSet) Copy() *ExtendedSet {
+	return &ExtendedSet{
+		statefulConn:    e.statefulConn,
+		nonStatefulConn: e.nonStatefulConn,
+		otherConn:       e.otherConn,
+		conn:            e.conn,
+	}
+}
+
+func computeNonStatefulConn(conn, otherConn, statefulConn *connection.Set) *connection.Set {
+	return conn.Subtract(otherConn).Subtract(statefulConn)
+}
+
+func (e *ExtendedSet) Intersect(other *ExtendedSet) *ExtendedSet {
+	statefulConn := e.statefulConn.Intersect(other.statefulConn)
+	otherConn := e.otherConn.Intersect(other.otherConn)
+	conn := e.conn.Intersect(other.conn)
+	return &ExtendedSet{
+		statefulConn:    statefulConn,
+		nonStatefulConn: computeNonStatefulConn(conn, otherConn, statefulConn),
+		otherConn:       otherConn,
+		conn:            conn,
+	}
```

Comment:
its look like there are some repeating patterns of creating ExtendedSet.
I would consider defining a constructor for each pattern.
for example, here the constructor will get statefulConn, otherConn and conn, only.
computeNonStatefulConn() should be called inside the constructor

Reply:
right. done.

Reply:
for now, you have two constuctors:
1. NewStateWithStateful() (need to be renamed)
2. EmptyConnWithStateful()
there are place for more constractors:
3. &ConnWithStateful{statefulConn: newTCPSet(), nonStatefulConn: NoConns(),
		otherConn: connection.All().Subtract(newTCPSet()), conn: connection.All()}
4. 				tcpFraction, nonTCPFraction := partitionTCPNonTCP(conn)
				v.AllowedConnsCombinedStateful.updateAllowedStatefulConnsMap(src, dst,
					&ConnWithStateful{statefulConn: tcpFraction, otherConn: nonTCPFraction,
						nonStatefulConn: connection.None(), conn: conn})
5. maybe more...
6.

Reply:
1 renamed
3 and 4 - code refactored to use the existing constructor (I don't see the point in having a constructor for a single use; moreover, the existing constructor is good enough in the relevant places)